File: lua/problem_72.lua
local M = {}

function M.largest_value_path(labels, edges)
    local n = #labels
    local adj = {}; for i = 1, n do adj[i] = {} end
    local indegree = {}; for i = 1, n do indegree[i] = 0 end
    for _, e in ipairs(edges) do
        if e[1] == e[2] then return nil end
        adj[e[1]][#adj[e[1]]+1] = e[2]
        indegree[e[2]] = indegree[e[2]] + 1
    end
    local dp = {}
    for i = 1, n do dp[i] = {}; for c = 0, 25 do dp[i][c] = 0 end end
    local queue = {}
    for i = 1, n do if indegree[i] == 0 then queue[#queue+1] = i end end
    local visited, best = 0, 0
    local qi = 1
    while qi <= #queue do
        local node = queue[qi]; qi = qi + 1
        visited = visited + 1
        local c = string.byte(labels, node) - string.byte('A')
        dp[node][c] = dp[node][c] + 1
        for cc = 0, 25 do if dp[node][cc] > best then best = dp[node][cc] end end
        for _, nb in ipairs(adj[node]) do
            for cc = 0, 25 do
                if dp[node][cc] > dp[nb][cc] then dp[nb][cc] = dp[node][cc] end
            end
            indegree[nb] = indegree[nb] - 1
            if indegree[nb] == 0 then queue[#queue+1] = nb end
        end
    end
    if visited ~= n then return nil end
    return best
end

function M.problem72()
    print("Problem 72:")
    assert(M.largest_value_path("ABACA", {{1,2},{1,3},{3,4},{4,5}}) == 3)
    assert(M.largest_value_path("A", {{1,1}}) == nil)
    print("All Tests Passed")
end

return M

File: c/problem_72.c
#include "problem_72.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

int largest_value_path(const char *labels, int n, int edges[][2], int m) {
    int adj[16][16]={0}, indegree[16]={0};
    for(int i=0;i<m;i++){
        if(edges[i][0]==edges[i][1]) return -1;
        adj[edges[i][0]][edges[i][1]]=1; indegree[edges[i][1]]++;
    }
    int dp[16][26]={0};
    int queue[16], qs=0, qe=0;
    for(int i=0;i<n;i++) if(indegree[i]==0) queue[qe++]=i;
    int visited=0, best=0;
    while(qs<qe){
        int node=queue[qs++]; visited++;
        int c=labels[node]-'A';
        dp[node][c]++;
        for(int k=0;k<26;k++) if(dp[node][k]>best) best=dp[node][k];
        for(int nb=0;nb<n;nb++) if(adj[node][nb]){
            for(int k=0;k<26;k++) if(dp[node][k]>dp[nb][k]) dp[nb][k]=dp[node][k];
            if(--indegree[nb]==0) queue[qe++]=nb;
        }
    }
    return (visited==n)?best:-1;
}

void problem72(void) {
    printf("Problem 72:\n");
    int e1[4][2]={{0,1},{0,2},{2,3},{3,4}};
    assert(largest_value_path("ABACA",5,e1,4)==3);
    int e2[1][2]={{0,0}};
    assert(largest_value_path("A",1,e2,1)==-1);
    printf("All Tests Passed\n");
}
